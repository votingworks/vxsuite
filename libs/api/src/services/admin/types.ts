import {
  BallotLocale,
  BallotLocaleSchema,
  BallotStyleId,
  BallotStyleIdSchema,
  CastVoteRecord,
  ContestId,
  ContestIdSchema,
  ContestOptionId,
  ContestOptionIdSchema,
  ElectionDefinition,
  ElectionDefinitionSchema,
  Id,
  IdSchema,
  Iso8601Timestamp,
  Iso8601TimestampSchema,
  PrecinctId,
  PrecinctIdSchema,
  Rect,
  RectSchema,
} from '@votingworks/types';
import * as z from 'zod';

/**
 * An election definition and associated DB metadata.
 */
export interface ElectionRecord {
  readonly id: Id;
  readonly electionDefinition: ElectionDefinition;
  readonly createdAt: Iso8601Timestamp;
  readonly isOfficialResults: boolean;
}

/**
 * Schema for {@link ElectionRecord}.
 */
export const ElectionRecordSchema: z.ZodSchema<ElectionRecord> = z.object({
  id: IdSchema,
  electionDefinition: ElectionDefinitionSchema,
  createdAt: Iso8601TimestampSchema,
  isOfficialResults: z.boolean(),
});

/**
 * Info related to a CVR file import attempt.
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type CvrFileImportInfo = {
  id: Id;
  alreadyPresent: number;
  exportedTimestamp: Iso8601Timestamp;
  fileMode: CvrFileMode;
  fileName: string;
  newlyAdded: number;
  scannerIds: string[];
  wasExistingFile: boolean;
};

/**
 * A cast vote record file's metadata.
 */
export interface CastVoteRecordFileRecord {
  readonly id: Id;
  readonly electionId: Id;
  readonly filename: string;
  readonly sha256Hash: string;
  readonly createdAt: Iso8601Timestamp;
}

/**
 * Schema for {@link CastVoteRecordFileRecord}.
 */
export const CastVoteRecordFileRecordSchema: z.ZodSchema<CastVoteRecordFileRecord> =
  z.object({
    id: IdSchema,
    electionId: IdSchema,
    filename: z.string().nonempty(),
    sha256Hash: z.string().nonempty(),
    createdAt: Iso8601TimestampSchema,
  });

/**
 * A Cast Vote Record's metadata.
 */
export interface CastVoteRecordFileEntryRecord {
  readonly id: Id;
  readonly electionId: Id;
  readonly data: string;
  readonly createdAt: Iso8601Timestamp;
}

/**
 * Schema for {@link CastVoteRecordFileEntryRecord}.
 */
export const CastVoteRecordFileEntryRecordSchema: z.ZodSchema<CastVoteRecordFileEntryRecord> =
  z.object({
    id: IdSchema,
    electionId: IdSchema,
    data: z.string(),
    createdAt: Iso8601TimestampSchema,
  });

/**
 * Status values for a write-in adjudication.
 */
export type WriteInAdjudicationStatus =
  | 'pending'
  | 'transcribed'
  | 'adjudicated';

/**
 * Schema for {@link WriteInAdjudicationStatus}.
 */
export const WriteInAdjudicationStatusSchema: z.ZodSchema<WriteInAdjudicationStatus> =
  z.union([
    z.literal('pending'),
    z.literal('transcribed'),
    z.literal('adjudicated'),
  ]);

/**
 * A write-in that has no transcription yet.
 */
export interface WriteInRecordPendingTranscription {
  readonly id: Id;
  readonly contestId: ContestId;
  readonly optionId: ContestOptionId;
  readonly castVoteRecordId: Id;
  readonly status: 'pending';
}

/**
 * Schema for {@link WriteInRecordPendingTranscription}.
 */
export const WriteInRecordPendingTranscriptionSchema: z.ZodSchema<WriteInRecordPendingTranscription> =
  z.object({
    id: IdSchema,
    contestId: ContestIdSchema,
    optionId: ContestOptionIdSchema,
    castVoteRecordId: IdSchema,
    status: z.literal('pending'),
  });

/**
 * A write-in that has a transcription but no adjudication yet.
 */
export interface WriteInRecordTranscribed {
  readonly id: Id;
  readonly contestId: ContestId;
  readonly optionId: ContestOptionId;
  readonly castVoteRecordId: Id;
  readonly status: 'transcribed';
  readonly transcribedValue: string;
}

/**
 * Schema for {@link WriteInRecordTranscribed}.
 */
export const WriteInRecordTranscribedSchema: z.ZodSchema<WriteInRecordTranscribed> =
  z.object({
    id: IdSchema,
    contestId: ContestIdSchema,
    optionId: ContestOptionIdSchema,
    castVoteRecordId: IdSchema,
    status: z.literal('transcribed'),
    transcribedValue: z.string().nonempty(),
  });

/**
 * A write-in that has been adjudicated.
 */
export interface WriteInRecordAdjudicated {
  readonly id: Id;
  readonly contestId: ContestId;
  readonly optionId: ContestOptionId;
  readonly castVoteRecordId: Id;
  readonly status: WriteInAdjudicationStatus;
  readonly transcribedValue: string;
  readonly adjudicatedValue: string;
  readonly adjudicatedOptionId?: ContestOptionId;
}

/**
 * Schema for {@link WriteInRecordAdjudicated}.
 */
export const WriteInRecordAdjudicatedSchema: z.ZodSchema<WriteInRecordAdjudicated> =
  z.object({
    id: IdSchema,
    contestId: ContestIdSchema,
    optionId: ContestOptionIdSchema,
    castVoteRecordId: IdSchema,
    status: z.literal('transcribed'),
    transcribedValue: z.string().nonempty(),
    adjudicatedValue: z.string().nonempty(),
    adjudicatedOptionId: ContestOptionIdSchema.optional(),
  });

/**
 * Information about a write-in in one of the adjudication states.
 */
export type WriteInRecord =
  | WriteInRecordPendingTranscription
  | WriteInRecordTranscribed
  | WriteInRecordAdjudicated;

/**
 * Schema for {@link WriteInRecord}.
 */
export const WriteInsRecordSchema: z.ZodSchema<WriteInRecord> = z.union([
  WriteInRecordPendingTranscriptionSchema,
  WriteInRecordTranscribedSchema,
  WriteInRecordAdjudicatedSchema,
]);

/**
 * Write-in adjudication information.
 */
export interface WriteInAdjudicationRecord {
  readonly id: Id;
  readonly contestId: ContestId;
  readonly transcribedValue: string;
  readonly adjudicatedValue: string;
  readonly adjudicatedOptionId?: ContestOptionId;
}

/**
 * Schema for {@link WriteInAdjudicationRecord}.
 */
export const WriteInAdjudicationRecordSchema: z.ZodSchema<WriteInAdjudicationRecord> =
  z.object({
    id: IdSchema,
    contestId: ContestIdSchema,
    transcribedValue: z.string().nonempty(),
    adjudicatedValue: z.string().nonempty(),
    adjudicatedOptionId: ContestOptionIdSchema.optional(),
  });

/**
 * Write-in summary information for write-ins pending transcription.
 */
export interface WriteInSummaryEntryPendingTranscription {
  readonly status: 'pending';
  readonly contestId: ContestId;
  readonly writeInCount: number;
}

/**
 * Schema for {@link WriteInSummaryEntryPendingTranscription}.
 */
export const WriteInSummaryEntryPendingTranscriptionSchema: z.ZodSchema<WriteInSummaryEntryPendingTranscription> =
  z.object({
    status: z.literal('pending'),
    contestId: ContestIdSchema,
    writeInCount: z.number().int().nonnegative(),
  });

/**
 * Write-in summary information for transcribed write-ins.
 */
export interface WriteInSummaryEntryTranscribed {
  readonly status: 'transcribed';
  readonly contestId: ContestId;
  readonly writeInCount: number;
  readonly transcribedValue: string;
}

/**
 * Schema for {@link WriteInSummaryEntryTranscribed}.
 */
export const WriteInSummaryEntryTranscribedSchema: z.ZodSchema<WriteInSummaryEntryTranscribed> =
  z.object({
    status: z.literal('transcribed'),
    contestId: ContestIdSchema,
    writeInCount: z.number().int().nonnegative(),
    transcribedValue: z.string().nonempty(),
  });

/**
 * Write-in summary information for adjudicated write-ins.
 */
export interface WriteInSummaryEntryAdjudicated {
  readonly status: 'adjudicated';
  readonly contestId: ContestId;
  readonly writeInCount: number;
  readonly transcribedValue: string;
  readonly writeInAdjudication: WriteInAdjudicationRecord;
}

/**
 * Schema for {@link WriteInSummaryEntryAdjudicated}.
 */
export const WriteInSummaryEntryAdjudicatedSchema: z.ZodSchema<WriteInSummaryEntryAdjudicated> =
  z.object({
    status: z.literal('adjudicated'),
    contestId: ContestIdSchema,
    writeInCount: z.number().int().nonnegative(),
    transcribedValue: z.string().nonempty(),
    writeInAdjudication: WriteInAdjudicationRecordSchema,
  });

/**
 * Write-in summary information.
 */
export type WriteInSummaryEntry =
  | WriteInSummaryEntryPendingTranscription
  | WriteInSummaryEntryTranscribed
  | WriteInSummaryEntryAdjudicated;

/**
 * Schema for {@link WriteInSummaryEntry}.
 */
export const WriteInSummaryEntrySchema: z.ZodSchema<WriteInSummaryEntry> =
  z.union([
    WriteInSummaryEntryPendingTranscriptionSchema,
    WriteInSummaryEntryTranscribedSchema,
    WriteInSummaryEntryAdjudicatedSchema,
  ]);

/**
 * An option for selecting an adjudication value in the write-in adjudication
 * table.
 */
export interface WriteInAdjudicationTableOption {
  readonly adjudicatedValue: string;
  readonly adjudicatedOptionId?: ContestOptionId;
  readonly enabled: boolean;
}

/**
 * Schema for {@link WriteInAdjudicationTableOption}.
 */
export const WriteInAdjudicationTableOptionSchema: z.ZodSchema<WriteInAdjudicationTableOption> =
  z.object({
    adjudicatedValue: z.string().nonempty(),
    adjudicatedOptionId: ContestOptionIdSchema.optional(),
    enabled: z.boolean(),
  });

/**
 * An option group for selecting an adjudication value in the write-in
 * adjudication table.
 */
export interface WriteInAdjudicationTableOptionGroup {
  readonly title: string;
  readonly options: readonly WriteInAdjudicationTableOption[];
}

/**
 * Schema for {@link WriteInAdjudicationTableOptionGroup}.
 */
export const WriteInAdjudicationTableOptionGroupSchema: z.ZodSchema<WriteInAdjudicationTableOptionGroup> =
  z.object({
    title: z.string().nonempty(),
    options: z.array(WriteInAdjudicationTableOptionSchema),
  });

/**
 * A row in the write-in adjudication table that has already been adjudicated.
 */
export interface WriteInAdjudicationTableAdjudicatedRow {
  readonly writeInCount: number;
  readonly transcribedValue: string;
  readonly writeInAdjudicationId: Id;
  readonly editable: boolean;
  readonly adjudicationOptionGroups: readonly WriteInAdjudicationTableOptionGroup[];
}

/**
 * Schema for {@link WriteInAdjudicationTableAdjudicatedRow}.
 */
export const WriteInAdjudicationTableAdjudicatedRowSchema: z.ZodSchema<WriteInAdjudicationTableAdjudicatedRow> =
  z.object({
    writeInCount: z.number().int().nonnegative(),
    transcribedValue: z.string().nonempty(),
    writeInAdjudicationId: IdSchema,
    editable: z.boolean(),
    adjudicationOptionGroups: z.array(
      WriteInAdjudicationTableOptionGroupSchema
    ),
  });

/**
 * Group of rows in the write-in adjudication table that have already been
 * adjudicated.
 */
export interface WriteInAdjudicationTableAdjudicatedRowGroup {
  readonly writeInCount: number;
  readonly adjudicatedValue: string;
  readonly adjudicatedOptionId?: ContestOptionId;
  readonly rows: readonly WriteInAdjudicationTableAdjudicatedRow[];
}

/**
 * Schema for {@link WriteInAdjudicationTableAdjudicatedRowGroup}.
 */
export const WriteInAdjudicationTableAdjudicatedRowGroupSchema: z.ZodSchema<WriteInAdjudicationTableAdjudicatedRowGroup> =
  z.object({
    writeInCount: z.number().int().nonnegative(),
    adjudicatedValue: z.string().nonempty(),
    adjudicatedOptionId: ContestOptionIdSchema.optional(),
    rows: z.array(WriteInAdjudicationTableAdjudicatedRowSchema),
  });

/**
 * A row in the write-in adjudication table that has not yet been adjudicated.
 */
export interface WriteInAdjudicationTableTranscribedRow {
  readonly writeInCount: number;
  readonly transcribedValue: string;
  readonly adjudicationOptionGroups: readonly WriteInAdjudicationTableOptionGroup[];
}

/**
 * Schema for {@link WriteInAdjudicationTableTranscribedRow}.
 */
export const WriteInAdjudicationTableTranscribedRowSchema: z.ZodSchema<WriteInAdjudicationTableTranscribedRow> =
  z.object({
    writeInCount: z.number().int().nonnegative(),
    transcribedValue: z.string().nonempty(),
    adjudicationOptionGroups: z.array(
      WriteInAdjudicationTableOptionGroupSchema
    ),
  });

/**
 * Group of rows in the write-in adjudication table that have not yet been
 * adjudicated.
 */
export interface WriteInAdjudicationTableTranscribedRowGroup {
  readonly writeInCount: number;
  readonly rows: readonly WriteInAdjudicationTableTranscribedRow[];
}

/**
 * Schema for {@link WriteInAdjudicationTableTranscribedRowGroup}.
 */
export const WriteInAdjudicationTableTranscribedRowGroupSchema: z.ZodSchema<WriteInAdjudicationTableTranscribedRowGroup> =
  z.object({
    writeInCount: z.number().int().nonnegative(),
    rows: z.array(WriteInAdjudicationTableTranscribedRowSchema),
  });

/**
 * Write-in adjudication table information, for use in adjudicating write-ins.
 */
export interface WriteInAdjudicationTable {
  readonly contestId: ContestId;
  readonly writeInCount: number;
  readonly adjudicated: readonly WriteInAdjudicationTableAdjudicatedRowGroup[];
  readonly transcribed: WriteInAdjudicationTableTranscribedRowGroup;
}

/**
 * Schema for {@link WriteInAdjudicationTable}.
 */
export const WriteInAdjudicationTableSchema: z.ZodSchema<WriteInAdjudicationTable> =
  z.object({
    contestId: ContestIdSchema,
    writeInCount: z.number().int().nonnegative(),
    adjudicated: z.array(WriteInAdjudicationTableAdjudicatedRowGroupSchema),
    transcribed: WriteInAdjudicationTableTranscribedRowGroupSchema,
  });

/**
 * A non-pending write-in summary entry.
 */
export type WriteInSummaryEntryNonPending =
  | WriteInSummaryEntryTranscribed
  | WriteInSummaryEntryAdjudicated;

/**
 * Schema for {@link WriteInSummaryEntryNonPending}.
 */
export const WriteInSummaryEntryNonPendingSchema: z.ZodSchema<WriteInSummaryEntryNonPending> =
  z.union([
    WriteInSummaryEntryTranscribedSchema,
    WriteInSummaryEntryAdjudicatedSchema,
  ]);

/**
 * Write-in image information.
 */
export interface WriteInImageEntry {
  readonly image: string;
  readonly ballotCoordinates: Rect;
  readonly contestCoordinates: Rect;
  readonly writeInCoordinates: Rect;
}

/**
 * Schema for {@link WriteInImageEntry}.
 */
export const WriteInImageEntrySchema: z.ZodSchema<WriteInImageEntry> = z.object(
  {
    image: z.string().nonempty(),
    ballotCoordinates: RectSchema,
    contestCoordinates: RectSchema,
    writeInCoordinates: RectSchema,
  }
);

/**
 * Cast vote record data for a given write in option.
 */
export interface CastVoteRecordData {
  readonly cvr: CastVoteRecord;
  readonly writeInId: Id;
  readonly contestId: ContestId;
  readonly optionId: ContestOptionId;
  readonly electionId: Id;
}

/**
 * Convenience enum for the printable ballot types.
 */
export const PrintableBallotType = {
  Absentee: 'absentee',
  Precinct: 'standard',
} as const;

/**
 * Printable ballot types.
 */
export type PrintableBallotType =
  typeof PrintableBallotType[keyof typeof PrintableBallotType];

/**
 * Schema for {@link PrintableBallotType}.
 */
export const PrintableBallotTypeSchema = z.union([
  z.literal('absentee'),
  z.literal('standard'),
]);

/**
 * Ballot mode.
 */
export enum BallotMode {
  /** Real ballots to be used and scanned during an election */
  Official = 'live',
  /** Test ballots to be used and scanned during pre-election testing / L&A */
  Test = 'test',
  /** Sample ballots to be provided to voters ahead of an election */
  Sample = 'sample',
  /** Draft ballots to verify that an election definition has been properly configured */
  Draft = 'draft',
}

/**
 * The ballot type for the CVR files currently being handled.
 */
export enum CvrFileMode {
  /** Only working with real CVR files generated during an official election. */
  Official = 'live',
  /** Only working with test CVR files used during pre-election testing/L&A. */
  Test = 'test',
  /** No CVR files imported yet - file mode is not currently locked. */
  Unlocked = 'unlocked',
}

/**
 * Schema for {@link BallotMode}.
 */
export const BallotModeSchema = z.nativeEnum(BallotMode);

/**
 * Information about printed ballots.
 */
export interface PrintedBallot {
  readonly ballotStyleId: BallotStyleId;
  readonly precinctId: PrecinctId;
  readonly locales: BallotLocale;
  readonly ballotType: PrintableBallotType;
  readonly ballotMode: BallotMode;
  readonly numCopies: number;
}

/**
 * Schema for {@link PrintedBallot}.
 */
export const PrintedBallotSchema: z.ZodSchema<PrintedBallot> = z.object({
  ballotStyleId: BallotStyleIdSchema,
  precinctId: PrecinctIdSchema,
  locales: BallotLocaleSchema,
  ballotType: PrintableBallotTypeSchema,
  ballotMode: BallotModeSchema,
  numCopies: z.number().int().nonnegative(),
});

/**
 * Database record for a printed ballot.
 */
export interface PrintedBallotRecord extends PrintedBallot {
  readonly id: Id;
  readonly electionId: Id;
  readonly createdAt: Iso8601Timestamp;
}

/**
 * Schema for {@link PrintedBallotRecord}.
 */
export const PrintedBallotRecordSchema: z.ZodSchema<PrintedBallotRecord> = z
  .object({
    id: IdSchema,
    electionId: IdSchema,
    ballotStyleId: BallotStyleIdSchema,
    precinctId: PrecinctIdSchema,
    locales: BallotLocaleSchema,
    ballotType: PrintableBallotTypeSchema,
    ballotMode: BallotModeSchema,
    numCopies: z.number().int().min(1),
    createdAt: Iso8601TimestampSchema,
  })
  .strict();
