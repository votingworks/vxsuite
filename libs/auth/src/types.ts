/**
 * Valid values for cert card type field
 */
export type CardType =
  | 'system-administrator'
  | 'election-manager'
  | 'poll-worker'
  | 'poll-worker-with-pin';

/**
 * Parsed custom cert fields
 */
export interface CustomCertFields {
  component: 'card' | 'admin' | 'central-scan' | 'mark' | 'scan';
  jurisdiction: string;
  cardType?: CardType;
  electionHash?: string;
}
