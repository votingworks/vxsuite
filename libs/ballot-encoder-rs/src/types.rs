use std::num::NonZeroU8;

use static_assertions::const_assert;

use crate::codable;

codable!(PrecinctIndex, u32, 0..=4096);
codable!(BallotStyleIndex, u32, 0..=4096);
codable!(PageNumber, u8, 1..=30);

// Statically validate our maximum values fit within the types we're using.
// TODO: move this into `codable!`
const_assert!(PrecinctIndex::BITS <= usize::BITS);
const_assert!(BallotStyleIndex::BITS <= usize::BITS);
const_assert!(PageNumber::BITS <= u8::BITS);

impl PageNumber {
    /// Whether this is the first page of its sheet, i.e. the front.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    ///
    /// assert!(page_one.is_recto());
    /// assert!(!page_two.is_recto());
    /// ```
    #[must_use]
    pub const fn is_recto(self) -> bool {
        self.0 % 2 == 1
    }

    /// Whether this is the second page of its sheet, i.e. the back.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    ///
    /// assert!(!page_one.is_verso());
    /// assert!(page_two.is_verso());
    /// ```
    #[must_use]
    pub const fn is_verso(self) -> bool {
        self.0 % 2 == 0
    }

    /// Determines the sheet number.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = PageNumber::new_unchecked(2);
    /// let page_three = PageNumber::new_unchecked(3);
    /// let page_four = PageNumber::new_unchecked(4);
    ///
    /// assert_eq!(page_one.sheet_number().get(), 1);
    /// assert_eq!(page_two.sheet_number().get(), 1);
    /// assert_eq!(page_three.sheet_number().get(), 2);
    /// assert_eq!(page_four.sheet_number().get(), 2);
    /// ```
    ///
    /// # Panics
    ///
    /// If `Self` is an invalid `PageNumber` that would produce a sheet number
    /// of zero, this method will panic.
    #[must_use]
    pub const fn sheet_number(self) -> NonZeroU8 {
        let verso_page = if self.is_verso() {
            self
        } else {
            self.opposite()
        };
        NonZeroU8::new(verso_page.get() / 2).expect("sheet number >= 1")
    }

    /// Gets the `PageNumber` opposite this one on the same sheet.
    ///
    /// ```
    /// # use ballot_encoder_rs::types::PageNumber;
    /// let page_one = PageNumber::new_unchecked(1);
    /// let page_two = page_one.opposite();
    /// assert_eq!(page_two.get(), 2);
    /// ```
    pub const fn opposite(self) -> Self {
        match self.0 % 2 {
            0 => Self(self.0 - 1),
            1 => Self(self.0 + 1),
            _ => unreachable!(),
        }
    }
}

#[cfg(test)]
pub mod tests {
    use std::num::NonZeroU8;

    use proptest::{prelude::Strategy, proptest};

    use super::PageNumber;

    #[test]
    fn test_page_number() {
        let page_one = PageNumber::new_unchecked(1);
        let page_two = PageNumber::new_unchecked(2);

        assert_eq!(page_one.get(), 1);
        assert_eq!(page_two.get(), 2);

        assert_eq!(page_one, page_one);
        assert_ne!(page_one, page_two);

        assert!(page_one.is_recto());
        assert!(!page_two.is_recto());
        assert!(!page_one.is_verso());
        assert!(page_two.is_verso());

        assert_eq!(page_one.sheet_number(), NonZeroU8::new(1).unwrap());
        assert_eq!(page_two.sheet_number(), NonZeroU8::new(1).unwrap());
    }

    pub fn arbitrary_page_number() -> impl Strategy<Value = PageNumber> {
        (PageNumber::MIN_VALUE..=PageNumber::MAX_VALUE).prop_map(PageNumber::new_unchecked)
    }

    proptest! {
        #[test]
        fn test_valid_page_number(page_number in arbitrary_page_number()) {
            assert_eq!(page_number, page_number.opposite().opposite());
            assert_eq!(page_number.is_recto(), page_number.opposite().is_verso());
            assert_eq!(page_number.is_verso(), page_number.opposite().is_recto());
            assert_eq!(page_number.sheet_number(), page_number.opposite().sheet_number());
        }

        #[test]
        fn test_invalid_page_number(value in PageNumber::MAX_VALUE + 1..=u8::MAX) {
            assert_eq!(PageNumber::new(value), None);
        }
    }
}
