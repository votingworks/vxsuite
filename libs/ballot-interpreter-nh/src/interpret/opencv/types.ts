import * as cv from '@u4/opencv4nodejs';

import {
  GridLayout,
  GridPosition,
  HmpbBallotPageMetadata,
  PageInterpretation,
  Result,
} from '@votingworks/types';

/** TODO */
export interface PageImageData {
  binarizedImage: cv.Mat;
  optimizedImage: cv.Mat;
  originalImage: cv.Mat;
}

/** Single page interpretation result. */
export interface OpenCvInterpretResult {
  interpretation: PageInterpretation;
  normalizedImageFilePath: string;
  originalImageFilePath: string;
}

/** TODO */
export interface PageRegion {
  image: cv.Mat;
  offset: cv.Point2;
}

/** TODO */
export interface PageMargins<T> {
  bottom: T;
  left: T;
  right: T;
  top: T;
}

/** TODO */
export type MarginType = keyof PageMargins<unknown>;

/** TODO */
export interface PotentialTimingMarks {
  allContours: cv.Contour[];
  potentialMarks: cv.Contour[];
}

/** TODO */
export interface DetectedTimingMarks {
  marginImage: PageRegion;
  marginType: MarginType;
  marks: cv.Contour[];
  medianHeight: number;
  medianSpacing: number;
  medianWidth: number;
  // pageRegion: PageRegion;
}

/** TODO */
export interface TimingMarks {
  detected: cv.Contour[];
  first: cv.Rect;
  interpolated: cv.Rect[];
  last: cv.Rect;
  medianSpacing: number;
  marginType: MarginType;
}

/** TODO */
export type BallotMetadata = Pick<
  HmpbBallotPageMetadata,
  'ballotStyleId' | 'ballotType' | 'electionHash' | 'isTestMode' | 'precinctId'
> & {
  gridLayout: GridLayout;
};

/** TODO */
export interface OvalTemplate {
  image: cv.Mat;
  shadedPixelRatio: number;
}

/** TODO */
export interface OvalMark {
  readonly gridPosition: GridPosition;
  readonly expectedBounds: Result<cv.Rect, string>;
  readonly matchBounds: Result<cv.Rect, string>;
  readonly score: number;
  readonly searchBounds: Result<cv.Rect, string>;
}
