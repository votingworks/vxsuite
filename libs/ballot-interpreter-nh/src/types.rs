// Defines a new type that wraps a String for use as an ID.
macro_rules! idtype {
    ($name:ident) => {
        #[derive(Clone, Debug, PartialEq, Eq, Hash, serde::Serialize, serde::Deserialize)]
        pub struct $name(String);

        impl $name {
            #[allow(dead_code)]
            pub const fn from(s: String) -> Self {
                Self(s)
            }
        }

        impl std::fmt::Display for $name {
            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
                write!(f, "{}", self.0)
            }
        }
    };
}

pub(crate) use idtype;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_idtype() {
        idtype!(Foo);
        let foo = Foo::from("foo".to_string());
        assert_eq!(format!("{}", foo), "foo");
        assert_eq!(serde_json::to_string(&foo).unwrap(), r#""foo""#);
        assert_eq!(
            serde_json::from_str::<Foo>(r#""foo""#).unwrap(),
            Foo::from("foo".to_string())
        );
    }

    #[test]
    fn test_idtype_clone() {
        idtype!(Foo);
        let foo = Foo::from("foo".to_string());
        let foo2 = foo.clone();
        assert_eq!(foo, foo2);
    }
}
