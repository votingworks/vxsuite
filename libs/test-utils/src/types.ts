import { BoundFunctions, queries, RenderResult } from '@testing-library/react';

export type ReactTestingLibraryQueryable =
  | BoundFunctions<typeof queries>
  | RenderResult;
