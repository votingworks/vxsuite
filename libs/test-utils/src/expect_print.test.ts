import { ElementWithCallback, PrintOptions } from '@votingworks/types';
import React from 'react';
import {
  expectAllPrintsAsserted,
  expectPrint,
  ExpectPrintError,
  fakePrintElement,
  fakePrintElementWhenReady,
  resetExpectPrint,
} from './expect_print';

beforeEach(() => {
  resetExpectPrint();
});

function getElement(textContent: string): JSX.Element {
  return React.createElement('div', undefined, textContent);
}

function getElementWithCallback(textContent: string): ElementWithCallback {
  return (onReadyToPrint: VoidFunction) => {
    onReadyToPrint();
    return React.createElement('div', undefined, textContent);
  };
}

const simpleElement = getElement('simple');
const simpleElementWithCallback = getElementWithCallback('simple');
const fakeOptions: PrintOptions = { sides: 'one-sided' };

test('fakePrintElement functions throw errors on un-asserted elements', async () => {
  expect.assertions(4);

  try {
    await fakePrintElement(simpleElement, fakeOptions);
    await fakePrintElement(simpleElement, fakeOptions);
  } catch (e) {
    expect(e).toBeInstanceOf(ExpectPrintError);
  }

  try {
    await fakePrintElement(simpleElement, fakeOptions);
    await fakePrintElementWhenReady(simpleElementWithCallback, fakeOptions);
  } catch (e) {
    expect(e).toBeInstanceOf(ExpectPrintError);
  }

  try {
    await fakePrintElementWhenReady(simpleElementWithCallback, fakeOptions);
    await fakePrintElement(simpleElement, fakeOptions);
  } catch (e) {
    expect(e).toBeInstanceOf(ExpectPrintError);
  }

  try {
    await fakePrintElementWhenReady(simpleElementWithCallback, fakeOptions);
    await fakePrintElementWhenReady(simpleElementWithCallback, fakeOptions);
  } catch (e) {
    expect(e).toBeInstanceOf(ExpectPrintError);
  }
});

describe('expectAllPrintsAsserted', () => {
  test('does not throw if no unasserted prints', () => {
    expect(expectAllPrintsAsserted).not.toThrow(ExpectPrintError);
  });

  test('does throw if element printed', async () => {
    await fakePrintElement(simpleElement, fakeOptions);
    expect(expectAllPrintsAsserted).toThrow(ExpectPrintError);
  });

  test('does throw if element with callback printed', async () => {
    await fakePrintElementWhenReady(simpleElementWithCallback, fakeOptions);
    expect(expectAllPrintsAsserted).toThrow(ExpectPrintError);
  });
});

describe('expectPrint', () => {
  test('throws error if no prints', async () => {
    expect.assertions(1);
    try {
      await expectPrint();
    } catch (e) {
      expect(e).toBeInstanceOf(ExpectPrintError);
    }
  });

  test('can expect multiple prints and use inspection', async () => {
    await fakePrintElement(getElement('1'), fakeOptions);
    await expectPrint((printedElement) => {
      expect(printedElement.getByText('1')).toBeTruthy();
    });

    await fakePrintElementWhenReady(getElementWithCallback('2'), fakeOptions);
    await expectPrint((printedElement) => {
      expect(printedElement.getByText('2')).toBeTruthy();
    });

    await fakePrintElementWhenReady(getElementWithCallback('3'), fakeOptions);
    await expectPrint((printedElement) => {
      expect(printedElement.getByText('3')).toBeTruthy();
    });

    await fakePrintElement(getElement('4'), fakeOptions);
    await expectPrint((printedElement) => {
      expect(printedElement.getByText('4')).toBeTruthy();
    });

    expectAllPrintsAsserted();
  });

  test('can expect prints without inspection', async () => {
    await fakePrintElement(simpleElement, fakeOptions);
    await expectPrint();

    await fakePrintElementWhenReady(simpleElementWithCallback, fakeOptions);
    await expectPrint();

    expectAllPrintsAsserted();
  });

  test('can access printOptions', async () => {
    expect.assertions(1);
    await fakePrintElement(simpleElement, { sides: 'two-sided-long-edge' });
    await expectPrint((_printedElement, printOptions) => {
      expect(printOptions).toMatchObject({ sides: 'two-sided-long-edge' });
    });
  });
});
