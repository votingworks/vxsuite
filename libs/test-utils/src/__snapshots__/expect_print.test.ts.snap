// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expectPrintToMatchSnapshot 1`] = `
<div>
  <div>
    simple
  </div>
</div>
`;
